prov/gni/src/gnix_datagram.c

```diff
@@ -277,8 +373,97 @@ int _gnix_dgram_bnd_post(struct gnix_datagram *d, gni_return_t *status_ptr)
 	return ret;
 }
 
+int  _gnix_dgram_poll(struct gnix_dgram_hndl *hndl,
+			enum gnix_dgram_poll_type type)
+{
+	int ret = FI_SUCCESS;
+	gni_return_t status;
+	gni_post_state_t post_state = GNI_POST_PENDING;
+	uint32_t responding_remote_id;
+	unsigned int responding_remote_addr;
+	struct gnix_datagram *dg_ptr;
+	uint64_t datagram_id = 0UL;
+	struct gnix_cm_nic *cm_nic = NULL;
+	struct gnix_address responding_addr;
+
+	cm_nic = hndl->cm_nic;
+	assert(cm_nic != NULL);
+
+	if (type == GNIX_DGRAM_BLOCK) {
+		status = GNI_PostdataProbeWaitById(cm_nic->gni_nic_hndl,
+						   -1,
+						   &datagram_id);
+		if ((status != GNI_RC_SUCCESS) &&
+			(status  != GNI_RC_TIMEOUT)) {
+			ret = gnixu_to_fi_errno(status);
+			goto err;
+		}
+	} else {
+		status = GNI_PostDataProbeById(cm_nic->gni_nic_hndl,
+						   &datagram_id);
+		if ((status != GNI_RC_SUCCESS) &&
+			(status  != GNI_RC_NO_MATCH)) {
+			ret = gnixu_to_fi_errno(status);
+			goto err;
+		}
+	}
+
+	if (status == GNI_RC_SUCCESS) {
+
+		dg_ptr = (struct gnix_datagram *)datagram_id;
+		assert(dg_ptr != NULL);
+
+		assert((dg_ptr->state == GNIX_DGRAM_STATE_CONNECTING) ||
+			(dg_ptr->state = GNIX_DGRAM_STATE_LISTENING));
+
+		/*
+		 * do need to take lock here
+		 */
+		fastlock_acquire(&cm_nic->lock);
+		status = GNI_EpPostDataTestById(dg_ptr->gni_ep,
+						datagram_id,
+						&post_state,
+						&responding_remote_addr,
+						&responding_remote_id);
+		fastlock_release(&cm_nic->lock);
+		if (status != GNI_RC_SUCCESS) {
+			ret = gnixu_to_fi_errno(status);
+			goto err;
+		}
+
+		switch (post_state) {
+		case GNI_POST_COMPLETED:
+			if (dg_ptr->callback_fn != NULL) {
+				responding_addr.device_addr =
+					responding_remote_addr;
+				responding_addr.cdm_id =
+					responding_remote_id;
+				ret = dg_ptr->callback_fn((void *)datagram_id,
+							responding_addr,
+							post_state);
+			}
+			break;
+		case GNI_POST_TIMEOUT:
+		case GNI_POST_TERMINATED:
+		case GNI_POST_ERROR:
```

Comment:
we think alike, I want to change this behavior in a subsequent PR and past the post_state to the callback for all four cases completec, timeout, terminated, error.  I could change this as part of this PR though.